### PR TITLE
feat(ingest): add `--log-file` option and show CLI logs in UI report

### PR DESCRIPTION
**Pull request by hsheth2**
- Add a `--log-file` attribute that writes debug logs and stdout (e.g. the final ingestion report) to a file. This can even be used to dump debug logs to a file while printing info logs to the terminal
- Includes the cli logs in the UI ingestion report. This currently matches the user's logging level for the cli run, but we can change that if we want to
- Colorize logs based on severity in the CLI (unless stderr is redirected to a file)
- Refactor logging setup

image

image

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by treff7es — APPROVED** [review]
Lgtm

File: metadata-ingestion/setup.py [hunk 1]
```diff
@@ -122,6 +122,8 @@ def get_long_description():
     "greenlet",
 }
 
+sqllineage_lib = "sqllineage==1.3.6"
```
Comment (jjoyce0510) [inline metadata-ingestion/setup.py:125]: nice refactor!

File: metadata-ingestion/src/datahub/cli/ingest_cli.py [hunk 2]
```diff
@@ -393,5 +393,5 @@ def rollback(
                     writer.writerow([row.get("urn")])
 
         except IOError as e:
-            print(e)
+            logger.exception(f"Unable to save rollback failure report: {e}")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/cli/ingest_cli.py:396]: Thank you

File: metadata-ingestion/src/datahub/cli/migrate.py [hunk 3]
```diff
@@ -262,7 +262,7 @@ def dataplatform2instance_func(
             delete_cli._delete_one_urn(src_entity_urn, soft=not hard, run_id=run_id)
         migration_report.on_entity_migrated(src_entity_urn, "status")  # type: ignore
 
-    print(f"{migration_report}")
+    click.echo(f"{migration_report}")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/cli/migrate.py:265]: omg thank you

File: metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py [hunk 4]
```diff
@@ -200,18 +195,29 @@ def on_completion(
...
+            type="CLI_INGEST",
+            serializedValue=structured_report_str,
+            contentType="application/json",
+        )
         execution_result_aspect = ExecutionRequestResultClass(
             status=status,
             startTimeMs=self.start_time_ms,
             durationMs=self.get_cur_time_in_ms() - self.start_time_ms,
-            report=json.dumps(report, indent=2),
+            report=summary,
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py:214]: Nice! So in this case, we are now sending the structured report to the correct field. From here, we can continue to add to that report and get it into our result aspect

Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py:214]: (Sets ourselves up for much richer reports)

File: metadata-ingestion/src/datahub/utilities/logging_manager.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        top_module = record.name.split(".")[0]
+
+        if top_module in DATAHUB_PACKAGES:
+            if self.debug:
+                return record.levelno >= logging.DEBUG
+            else:
+                return record.levelno >= logging.INFO
+        else:
+            if self.debug:
+                return record.levelno >= logging.WARNING
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/utilities/logging_manager.py:84]: nice. good idea to filter out non-datahub noise

File: metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py [hunk 6]
```diff
@@ -200,18 +195,29 @@ def on_completion(
         report: Dict[str, Any],
         ctx: PipelineContext,
     ) -> None:
+        # Prepare a nicely formatted summary
+        structured_report_str = json.dumps(report, indent=2)
+        summary = f"~~~~ Ingestion Report ~~~~\n{structured_report_str}\n\n"
+        summary += "~~~~ Ingestion Logs ~~~~\n"
+        summary += get_log_buffer().format_lines()
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py:202]: ah so here we leverage the in memory buffer. nice

File: metadata-ingestion/src/datahub/utilities/logging_manager.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        handlers = [
+            stream_handler,
+            file_handler,
+            buffer_handler,
+        ]
+
+        # Configure the loggers.
+        root_logger = logging.getLogger()
+        _remove_all_handlers(root_logger)
+        root_logger.setLevel(logging.INFO)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/utilities/logging_manager.py:171]: why do we need this "normalization" step?

File: metadata-ingestion/src/datahub/utilities/logging_manager.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            root_logger.addHandler(handler)
+
+        for lib in DATAHUB_PACKAGES:
+            # Using a separate handler from the root logger allows us to control the log level
+            # of the datahub libs independently of the root logger.
+            # It also insulates us from rogue libraries that might call `logging.basicConfig`
+            # or otherwise mess with the logging configuration.
+            lib_logger = logging.getLogger(lib)
+            _remove_all_handlers(lib_logger)
+            lib_logger.setLevel(logging.DEBUG)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/utilities/logging_manager.py:182]: why this step?

File: metadata-ingestion/src/datahub/utilities/tee_io.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from typing import Any, TextIO
+
+
+class TeeIO:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/utilities/tee_io.py:4]: Fancy!

File: metadata-ingestion/src/datahub/utilities/logging_manager.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        # Create stdout handler.
+        stream_handler = logging.StreamHandler()
+        stream_handler.addFilter(_DatahubLogFilter(debug=debug))
+        stream_handler.setFormatter(_stream_formatter)
+
+        # Create file handler.
+        file_handler: logging.Handler
+        if log_file:
+            file = stack.enter_context(open(log_file, "w"))
+            tee = TeeIO(sys.stdout, file)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/utilities/logging_manager.py:148]: I'm curious - why is this Tee and context piece even required? I'd imagine we should be able to create a global handler that directs to a file without requiring this right?

File: metadata-ingestion/tests/unit/test_cli_logging.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    print("this is a print statement")
+
+    click.echo("this is a click.echo statement")
+
+    logger.debug("Example debug line")
+    logger.info("This is an %s message", "info")  # test string interpolation
+    logger.warning("This is a warning message")
+    logger.error("this is an error with no stack trace")
+    try:
+        1 / 0
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_cli_logging.py:25]: i like this

File: metadata-ingestion/tests/unit/test_cli_logging.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        datahub, ["--debug", "--log-file", str(log_file), "my-logging-fn"]
+    )
+    assert result.exit_code == 0
+
+    # The output should include the stdout and stderr, formatted as expected.
+    regex.match(
+        r"""\
+this is a print statement
+this is a click.echo statement
+[.+] DEBUG    .datahub.my_cli_module:\d+. - Example debug line
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_cli_logging.py:44]: Ha! This is pretty cool

File: metadata-ingestion/tests/unit/test_cli_logging.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+[.+] INFO     .datahub.my_cli_module:\d+. - This is an info message
+[.+] WARNING  .datahub.my_cli_module:\d+. - This is a warning message
+[.+] ERROR    .datahub.my_cli_module:\d+. - this is an error with no stack trace
+[.+] ERROR    .datahub.my_cli_module:\d+. - failed to divide by zero
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_cli_logging.py:48]: Why does everything start with [.+]?

File: metadata-ingestion/tests/unit/test_cli_logging.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    )
+
+    # The log file should match the output exactly.
+    log_file_output = log_file.read_text()
+    assert log_file_output == result.output
+
+    # The in-memory log buffer should contain the log messages.
+    # The first two lines are stdout, so we skip them.
+    expected_log_output = "\n".join(result.output.splitlines()[2:])
+    assert get_log_buffer().format_lines() == expected_log_output
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_cli_logging.py:64]: Do we need to cleanup the file here?

File: metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py (hunk 4 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py:214]: yup exactly - this way the structured report shows up in the correct field for both managed ingestion and cli ingestion

note that the structured report field is basically a string right now - we may need to rethink that in the future

File: metadata-ingestion/src/datahub/utilities/logging_manager.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+
+_log_buffer = _LogBuffer(maxlen=IN_MEMORY_LOG_BUFFER_SIZE)
+
+
+def get_log_buffer() -> _LogBuffer:
+    return _log_buffer
+
+
+_stream_formatter = _ColorLogFormatter()
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/logging_manager.py:130]: yup added support for a `NO_COLOR` env flag

File: metadata-ingestion/src/datahub/utilities/logging_manager.py (hunk 10 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/logging_manager.py:148]: we're redirecting stdout here (e.g. click.echo and print statements)

we want those messages to show up both in the terminal and in the log_file, which is why we need the tee

File: metadata-ingestion/src/datahub/utilities/logging_manager.py (hunk 7 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/logging_manager.py:171]: the default is warning globally, and can also be messed with by other libraries

File: metadata-ingestion/tests/unit/test_cli_logging.py (hunk 13 above)
Comment (hsheth2) [reply metadata-ingestion/tests/unit/test_cli_logging.py:48]: to handle the timestamp part of the log

File: metadata-ingestion/tests/unit/test_cli_logging.py (hunk 14 above)
Comment (hsheth2) [reply metadata-ingestion/tests/unit/test_cli_logging.py:64]: nope its a pytest tmp_file so it gets cleaned up automatically

Comment (jjoyce0510) [reply metadata-ingestion/tests/unit/test_cli_logging.py:64]: nice!

**Review by jjoyce0510 — APPROVED** [review]
LGTM

**Pull request merged**